File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/execution/batch/BatchJobWatcher.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.core.implementation.execution.batch;

import java.util.LinkedList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.core.model.execution.batch.IBatchEnvironment;
import org.openmole.core.model.execution.batch.IBatchExecutionJob;
import org.openmole.core.model.job.IJob;
import org.openmole.core.implementation.internal.Activator;
import org.openmole.core.model.execution.IExecutionJobRegistry;
import org.openmole.misc.updater.IUpdatable;
import org.openmole.misc.workspace.ConfigurationLocation;

public class BatchJobWatcher implements IUpdatable {

    final static ConfigurationLocation CheckInterval = new ConfigurationLocation(BatchJobWatcher.class.getName(), "CheckInterval");

    static {
        Activator.getWorkspace().addToConfigurations(CheckInterval, "PT2M");
    }

    final IBatchEnvironment<?> watchedEnv;

    public BatchJobWatcher(IBatchEnvironment<?> watchedEnv) throws InternalProcessingError {
        super();
        this.watchedEnv = watchedEnv;
    }

    @Override
    public boolean update() throws InterruptedException {
        IExecutionJobRegistry<IBatchExecutionJob> registry = watchedEnv.getJobRegistry();
        List<IJob> jobGroupsToRemove = new LinkedList<IJob>();
        
        synchronized (registry) {
            for (IJob job : registry.getAllJobs()) {
         
                if (job.allMoleJobsFinished()) {
                    for (final IBatchExecutionJob ej : registry.getExecutionJobsFor(job)) {
                        ej.kill();
                    }

                    jobGroupsToRemove.add(job);
                } else {

                    List<IBatchExecutionJob<?>> executionJobsToRemove = new LinkedList<IBatchExecutionJob<?>>();

                    for (final IBatchExecutionJob<?> ej : registry.getExecutionJobsFor(job)) {
                        switch (ej.getState()) {
                            case KILLED:
                                executionJobsToRemove.add(ej);
                                break;
                        }
                    }

                    for (IBatchExecutionJob<?> ej : executionJobsToRemove) {
                        registry.remove(ej);
                    }

                    if (registry.getNbExecutionJobsForJob(job) == 0) {
                        try {
                            watchedEnv.submit(job);
                        } catch (InternalProcessingError e) {
                            Logger.getLogger(BatchJobWatcher.class.getName()).log(Level.SEVERE, "Submission of job failed, job isn't being executed.", e);
                        } catch (UserBadDataError e) {
                            Logger.getLogger(BatchJobWatcher.class.getName()).log(Level.SEVERE, "Submission of job failed, job isn't being executed.", e);
                        }
                    }
                }
            }

            for (IJob j : jobGroupsToRemove) {
                registry.removeJob(j);
            }
        }
        return true;
    }
}

File: openmole/core/org.openmole.core.batch/src/main/scala/org/openmole/core/batch/environment/BatchJobWatcher.scala
/*
 * Copyright (C) 2010 Romain Reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.batch.environment

import org.openmole.core.tools.service.Logger
import org.openmole.core.updater.IUpdatableWithVariableDelay
import org.openmole.core.workflow.execution.ExecutionState._
import org.openmole.core.workflow.job.Job
import org.openmole.core.workspace.Workspace
import collection.mutable._
import org.openmole.core.batch.refresh.{ Kill, Manage }
import scala.ref.WeakReference

object BatchJobWatcher extends Logger {
  case object Watch

  class ExecutionJobRegistry {
    val jobs = new HashMap[Job, Set[BatchExecutionJob]] with MultiMap[Job, BatchExecutionJob]

    def allJobs = jobs.keySet

    def executionJobs(job: Job) = jobs.getOrElse(job, Set.empty)

    def remove(ejob: BatchExecutionJob) = jobs.removeBinding(ejob.job, ejob)

    def isEmpty: Boolean = jobs.isEmpty

    def register(ejob: BatchExecutionJob) = jobs.addBinding(ejob.job, ejob)

    def removeJob(job: Job) = jobs -= job

    def allExecutionJobs = jobs.values.flatMap(_.toSeq)
  }

}

class BatchJobWatcher(environment: WeakReference[BatchEnvironment]) extends IUpdatableWithVariableDelay {

  val registry = new BatchJobWatcher.ExecutionJobRegistry

  import BatchJobWatcher._

  def register(job: BatchExecutionJob) = synchronized { registry.register(job) }

  override def update: Boolean = synchronized {
    val env = environment.get match {
      case None ⇒
        for (ej ← registry.allExecutionJobs) if (ej.state != KILLED) BatchEnvironment.jobManager ! Kill(ej)
        return false
      case Some(env) ⇒ env
    }

    val jobGroupsToRemove = new ListBuffer[Job]

    Log.logger.fine("Watch jobs " + registry.allJobs.size)
    for (job ← registry.allJobs) {
      if (job.finished) {
        for (ej ← registry.executionJobs(job)) if (ej.state != KILLED) BatchEnvironment.jobManager ! Kill(ej)
        jobGroupsToRemove += job
      }
      else {
        val executionJobsToRemove = new ListBuffer[BatchExecutionJob]

        for {
          ej ← registry.executionJobs(job)
          if ej.state.isFinal
        } executionJobsToRemove += ej

        for (ej ← executionJobsToRemove) registry.remove(ej)
        if (registry.executionJobs(job).isEmpty) env.submit(job)
      }
    }

    for (j ← jobGroupsToRemove) registry.removeJob(j)
    true
  }

  def executionJobs = synchronized { registry.allExecutionJobs }

  def delay = Workspace.preferenceAsDuration(BatchEnvironment.CheckInterval)
}
